service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/flight/RetrieveControlledBigQueryDatasetLifetimeStep.java

```diff
@@ -0,0 +1,134 @@
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys.CONTROLLED_BIG_QUERY_DATASETS_WITHOUT_LIFETIME;
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys.CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_PARTITION_LIFETIME_MAP;
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys.CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_TABLE_LIFETIME_MAP;
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys.CONTROLLED_RESOURCE_ID_TO_WORKSPACE_ID_MAP;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.resource.controlled.cloud.gcp.bqdataset.ControlledBigQueryDatasetResource;
+import bio.terra.workspace.service.resource.controlled.model.ControlledResource;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.api.services.bigquery.model.Dataset;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RetrieveControlledBigQueryDatasetLifetimeStep implements Step {
+  private static final Logger logger =
+      LoggerFactory.getLogger(RetrieveControlledBigQueryDatasetLifetimeStep.class);
+  private final CrlService crlService;
+
+  public RetrieveControlledBigQueryDatasetLifetimeStep(CrlService crlService) {
+    this.crlService = crlService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    validateRequiredEntries(
+        context.getWorkingMap(), CONTROLLED_BIG_QUERY_DATASETS_WITHOUT_LIFETIME);
+    List<ControlledBigQueryDatasetResource> controlledBigQueryDatasets =
+        context
+            .getWorkingMap()
+            .get(CONTROLLED_BIG_QUERY_DATASETS_WITHOUT_LIFETIME, new TypeReference<>() {});
+    Map<UUID, String> resourceIdToDefaultTableLifetimeMap = new HashMap<>();
+    Map<UUID, String> resourceIdToDefaultPartitionLifetimeMap = new HashMap<>();
+    Map<UUID, String> resourceIdToWorkspaceIdMap = new HashMap<>();
+
+    assert controlledBigQueryDatasets != null;
+    for (var resource : controlledBigQueryDatasets) {
+      logger.info(
+          "Getting default table lifetime and partition life for resource (BigQuery dataset) {} in workspace {}",
+          resource.getResourceId(),
+          resource.getWorkspaceId());
+      populateMapsWithResourceIdKey(
+          resourceIdToDefaultTableLifetimeMap,
+          resourceIdToDefaultPartitionLifetimeMap,
+          resourceIdToWorkspaceIdMap,
+          resource,
+          getBqDatasetDefaultTableLifetime(resource),
```

Comment:
This causes two cloud requests to lookup the same information. I think it would be better to make one request to get that, and then pull the two pieces of info from it.